Feature/multi az migration

jagadish-kb opened the pull request:
During multi-az migration, we would want to have two ‘Service Instance Update’ tasks. The first one must step-up the vms and the second one must step down the VMs or vice versa based on individual service preference.
With this feature serviceflows as a construct introduced into SF via which any of the existing granular operations could be tied together and performed as one single operation. (ex. Backup & Update - for major version upgrade, Restore & Backup for PITR etc..)
More info can be found in - HCPCFS-2034.

kamath-prasad commented on api-controllers/ServiceBrokerApiController.js:
```diff
@@ -134,36 +140,61 @@ class ServiceBrokerApiController extends FabrikBaseController {
-      .then(() => {
-        if (!plan.manager.async) {
-          return eventmesh.apiServerClient.getResourceOperationStatus({
+        serviceFlow = req._serviceFlow;
```
just for my understanding , in case a bi-weekly / scheduled update comes along with the multi-az update , then how do we ensure that the deployment resource is locked ? Since bi-weekly update is not via CF and in the multi-az update we do not lock the deployment resource till the task is initiated , we can have collision right ?

jagadish-kb replied:
The lock resource is created with the resource id of the  instance. The lock resource has options attrib (LockResourceDetails) which is used by lock manager to poll for checking the status and holding the lock. Since the instance id is the resource ID of the lock resource and it wont be released till the serialworkflow resource turns to success/failed, scheduled update wont pass through. This is similar to how locking is done for RESTORE or Backup operation.

kamath-prasad replied:
Ok , got it . thanks for the info.

jagadish-kb commented on operators/serviceflow-operator/SerialServiceFlowOperator.js:
```diff
@@ -0,0 +1,186 @@
+const apiServerClient = require('../../data-access-layer/eventmesh').apiServerClient;
+const BaseOperator = require('../BaseOperator');
+const TaskFabrik = require('./task/TaskFabrik');
+const FAILED = true;
```
reason to name it in assertion is because, its a constant and if it were to be named as suggested then there is a possibility of it being false as well. In this case, I want to signify a boolean which has failed always.

kamath-prasad replied:
ok ,can leave it as it is.

jagadish-kb commented on broker/config/crds/serviceflow.servicefabrik.io_v1alpha1_serialserviceflows.yaml:
```diff
@@ -0,0 +1,45 @@
+spec:
+  group: serviceflow.servicefabrik.io
+  names:
+    kind: Serialserviceflow
```
Yes. Thats how these CRDs are supposed to be written :-)

jagadish-kb commented on api-controllers/ServiceBrokerApiController.js:
```diff
@@ -218,14 +249,19 @@ class ServiceBrokerApiController extends FabrikBaseController {
 
     function done(result) {
       const body = _.pick(result, 'state', 'description');
+      if (body.state === CONST.APISERVER.RESOURCE_STATE.IN_PROGRESS) {
```
Yes.

kamath-prasad commented on operators/serviceflow-operator/serial-serviceflow-definition.yml:
```diff
@@ -0,0 +1,35 @@
+    task_description: 'Void blueprint task2'
+    task_data: 
+      sample: 'InputDataFrom Task2 ----'
+upgrade_to_multi_az:
```
one question , dont we also need another flow `downgrade_to_single_az` ?

amitmalav commented on operators/BaseOperator.js:
```diff
@@ -17,18 +17,20 @@ class BaseOperator {
     const queryString = `state in (${_.join(validStateList, ',')})`;
     logger.debug(`Registering watcher for resourceGroup ${resourceGroup} of type ${resourceType} with queryString as ${queryString}`);
-    return eventmesh.apiServerClient.registerWatcher(resourceGroup, resourceType, this.handleResource.bind(this), queryString)
+    return eventmesh.apiServerClient.registerWatcher(resourceGroup, resourceType, (resource) => this.handleResource(resource, handler, resourceGroup, resourceType, queryString), queryString)
```
Shall we rename `resource` to `changedObject` or `changedResource`?

jagadish-kb commented on operators/BaseOperator.js:
```diff
@@ -17,18 +17,20 @@ class BaseOperator {
     const queryString = `state in (${_.join(validStateList, ',')})`;
     logger.debug(`Registering watcher for resourceGroup ${resourceGroup} of type ${resourceType} with queryString as ${queryString}`);
-    return eventmesh.apiServerClient.registerWatcher(resourceGroup, resourceType, this.handleResource.bind(this), queryString)
+    return eventmesh.apiServerClient.registerWatcher(resourceGroup, resourceType, (resource) => this.handleResource(resource, handler, resourceGroup, resourceType, queryString), queryString)
```
I found adding this prefix 'changed' given, no need to specify it. i.e. the handlers are registered for change events as mentioned in query string, so if handlers are fired then obviously they are on change, so prefixing the obvious thing seems redundant. Also naming sth as 'object' which is too generic is a bad idea. Will stick with resource. Didnt change other parts of the code as it was existing.

jagadish-kb commented on operators/BaseOperator.js:
```diff
@@ -86,11 +89,10 @@ class BaseOperator {
   }
 
-  _preProcessRequest(objectBody, processingLockStatus) {
+  _preProcessRequest(objectBody, processingLockStatus, resourceGroup, resourceType, queryString) {
```
Debugging.

jagadish-kb commented on operators/serviceflow-operator/SerialServiceFlowOperator.js:
```diff
@@ -0,0 +1,186 @@
+    return Promise.try(() => {
+      assert.ok(resource.metadata.name, `Argument 'metadata.name' is required to run the task`);
+      assert.ok(resource.spec.options, `Argument 'spec.options' is required to run the task`);
+      const serviceFlowOptions = JSON.parse(resource.spec.options);
```
spec.options is mandatory. Not required to use lodash here, this would always pass. Look at LN 46.

jagadish-kb commented on operators/serviceflow-operator/StartSerialServiceFlow.js:
```diff
@@ -0,0 +1,8 @@
+
+const SerialServiceFlowOperator = require('./SerialServiceFlowOperator');
+const TaskOperator = require('./task/TaskOperator');
+const swfOperator = new SerialServiceFlowOperator();
```
earlier we used to call it workflow and this was renamed. Fixing it per suggestion.

sauravmndl commented on operators/serviceflow-operator/task/TaskOperator.js:
```diff
@@ -0,0 +1,130 @@
+            .continueToHoldLock(object)
+            .tap((resource) => {
+              object.metadata.resourceVersion = resource.metadata.resourceVersion;
+              logger.debug(`-> Retained the lock successfully -> for for ${object.metadata.name} with version - ${resource.metadata.resourceVersion}`);
```
double 'for'

jagadish-kb replied:
ok

sauravmndl commented on operators/serviceflow-operator/task/TaskOperator.js:
```diff
@@ -0,0 +1,130 @@
+          return task
+            .updateStatus(resourceDetails, status)
+            .then(() => this.clearPoller(object, intervalId))
+            .tap(() => logger.info('Released processing lock!!!'))
```
can we log id in statement here?

jagadish-kb replied:
ok

sauravmndl commented on broker/config/crds/serviceflow.servicefabrik.io_v1alpha1_serialserviceflows.yaml:
```diff
@@ -0,0 +1,45 @@
+spec:
+  group: serviceflow.servicefabrik.io
+  names:
+    kind: Serialserviceflow
```
Should it be in camel case ?

jagadish-kb replied:
Answered already. This how CRDs are supposed to be defined :-)

sauravmndl replied:
**# kind is normally the CamelCased singular type. Your resource manifests use this**
you can see some examples [here](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/)

jagadish-kb replied:
For some reason it was earlier throwing exeptions during development. Now changed to camelcase and it seems to work. @amitmalav  do you remember this weird issue with camel casing i was facing during development?

sauravmndl commented on operators/serviceflow-operator/task/TaskFabrik.js:
```diff
@@ -0,0 +1,31 @@
+const CONST = require('../../../common/constants');
+let ServiceInstanceUpdate, ServiceInstanceBackup, BlueprintTask;
+
+class TaskFabrik {
```
'FabrikTask' would have been better name in my opinion.....

jagadish-kb replied:
This follows the existing Fabrik naming conventions ex. `JobFabrik`. Will keep it as is, Fabrik, aka factory generally sounds better as a suffix than a prefix. Naming again, subjective :-)

sauravmndl commented:
Try to cover 100% for newly added lines.

jagadish-kb commented on api-controllers/FabrikBaseController.js:
```diff
@@ -35,12 +37,29 @@ class FabrikBaseController extends BaseController {
+                .tap(id => serviceFlowId = id);
+            }
+          }
+          return undefined;
```
wanted to be explicit & make it evident that service flow is undefined. Even not having a return has the same effect.

sauravmndl replied:
then we can return serviceFlowId which is undefined. Because returning undefined looks bit odd.

jagadish-kb replied:
Disagree here. Returning uninitialised serviceFlowId would be less readable as against the explicity `undefined` here. Not sure what is odd in returning `undefined` ? :-)

jagadish-kb commented on common/utils/ServiceFlowMapper.js:
```diff
@@ -0,0 +1,50 @@
+class ServiceFlowMapper {
+  getServiceFlow(params) {
+    let serviceFlow;
+    const flowCheckers = ['_checkForMultiAz', '_checkForBluePrint'];
```
No. These are function names within the same class. They are not resuable. Conveting them to consts would be overdoing and wont achieve the benefits of having constants truely. Also it would affect readability badly in this case.

jagadish-kb commented on common/utils/ServiceFlowMapper.js:
```diff
@@ -0,0 +1,50 @@
+  _checkForMultiAz(params) {
+    logger.debug(`Checking for multi-az-migrate - `);
+    if (_.get(params, 'parameters.multi_az') !== undefined) {
+      logger.info(`Multi-AZ_Upgrade Service Flow is to be executed`);
```
This method  does really know what else is there in the params object, which is quite generic. All this method cares about is params object having the attrib `multi_az` or not. So logging any id in this method would mean this method having idea of from where is this being called & accordingly prefix what the id would mean. One thing we could do is dump entire params object, but that would be too much clutter on log. For now will leave it as is, however if we really want to debug, then can add params object on this log stmt.

sauravmndl replied:
ok

jagadish-kb commented on common/utils/ServiceFlowMapper.js:
```diff
@@ -0,0 +1,50 @@
+    const currentPlanId = _.get(params, 'plan_id', '');
+    const prevPlanId = _.get(params, 'previous_values.plan_id', '');
+    if (currentPlanId !== prevPlanId) {
+      logger.info(`Plan has changed... This is being treated as Major Version upgrade for now...!`);
```
This method is not even a right method. See above at LN10, this is just a placeholder method and this strategy is not going to be evaluated anytime for now. This is just put in here for example, so for now will skip this comment, but yes when we add real strategy to identify major version upgrade, the loggers could be then enhanced with the above comment.

sauravmndl replied:
ok

jagadish-kb commented on data-access-layer/eventmesh/ApiServerClient.js:
```diff
@@ -547,6 +551,18 @@ class ApiServerClient {
+   * @param {string} opts.resourceType - Type of operation
+   * @param {string} opts.resourceId - Unique id of resource
+   */
+  getResourceStatus(opts) {
```
This wrapper method is on the lines of getLastOperation on APIServerClient, already defined. Anyways wrapper methods are no harm, they eliminate every caller having to do `_.get(resource, 'status')`.

jagadish-kb commented on api-controllers/FabrikBaseController.js:
```diff
@@ -49,24 +68,38 @@ class FabrikBaseController extends BaseController {
+  }
+
   _unlockIfReqfailed(operationType, processedRequest, lockId, req, res, next, err) {
     const plan_id = req.body.plan_id || req.query.plan_id;
```
existing code not modified in this PR. If one were to fix it truly, since plan_id is anyways required in the flow below, it must throw an BadRequest Error when its null.

jagadish-kb commented on data-access-layer/eventmesh/ApiServerClient.js:
```diff
@@ -170,11 +170,14 @@ class ApiServerClient {
   }
   parseResourceDetailsFromSelfLink(selfLink) {
     // self links are typically: /apis/deployment.servicefabrik.io/v1alpha1/namespaces/default/directors/d-7
```
sample link is the same right? this sample holds true for all resources. I hope you are not asking for a sample of serviceflow link?

kamath-prasad replied:
Sorry , my bad , I thought we added more details to the link . Now I see that its the same link , we just started accessing the item at 7th index.

jagadish-kb commented on operators/serviceflow-operator/task/Task.js:
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+
```
ok

jagadish-kb commented on operators/serviceflow-operator/SerialServiceFlowOperator.js:
```diff
@@ -0,0 +1,186 @@
+const apiServerClient = require('../../data-access-layer/eventmesh').apiServerClient;
+const BaseOperator = require('../BaseOperator');
+const TaskFabrik = require('./task/TaskFabrik');
+const FAILED = true;
```
It is used in LN177. Wanted a local boolean 'true' to be sent and to increase readability just added a local boolean constant FAILED.

jagadish-kb commented:
@sauravmndl  can u give example where u saw Code coverage is not 100% for new lines in this PR? I verified and it covers 100%. Please check code coverage report above by coveralls. It has increased overall coverage by 0.2%, this would not be possible if newly added lines are not 100% covered.

sauravmndl commented:
yesJagadish overall increased, just saying if we can cover 100% of modified/newly added lines.
**316 of 330 (95.76%)**

jagadish-kb commented:
@saurav - addressed coverage issues with regards to other files except BaseOperator as that was existing code and i had just modified loggers. Also its quite tricky to write test for those uncovered blocks. Also addressed other comments. 
Not sure how to run coveralls again on this.

ketakig12 commented on operators/serviceflow-operator/task/ServiceInstanceBackupTask.js:
```diff
@@ -0,0 +1,66 @@
+              lastOperation: {},
+              response: {}
+            }
+          })
```
Since the same createResource call is done from [startBackup](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/7a5417031e300e1accfd3284194a087b581b42ad/api-controllers/ServiceFabrikApiController.js#L343) as well, in future lets say we plan to add space_guid to labels, we will need to remember to change here as well?

ketakig12 replied:
I think it will be hard to refactor this code at common place, but then should we put a comment at both places?

jagadish-kb commented on operators/BaseOperator.js:
```diff
@@ -122,7 +127,7 @@ class BaseOperator {
       } else {
         processingLockStatus.conflict = true;
-        logger.info(`Request with options ${JSON.stringify(options)} is picked by process with ip ${lockedByManager} at ${processingStartedAt}`);
+        logger.debug(`Resource ${objectBody.metadata.name}  ${resourceGroup}, ${resourceType}, ${queryString} - is picked by process with ip ${lockedByManager} at ${processingStartedAt}`);
```
Generally if you open up the logs we see lot of such stmts which really dont add value since the watchers refresh at refresh interval and logs really get bloated with lots of such stmts which add no value per se. (i.e. we see xxx picked up by other host msgs a lot). So based on my experience I lowered it.

ketakig12 replied:
Sure, just thought that this might help in debugging which process picked up the request.

jagadish-kb commented on operators/BaseOperator.js:
```diff
@@ -17,18 +17,20 @@ class BaseOperator {
+          .delay(watchRefreshInterval)
           .then(() => {
-            logger.debug(`Refreshing stream after ${CONST.APISERVER.WATCHER_REFRESH_INTERVAL}`);
+            logger.debug(`Refreshing stream after ${watchRefreshInterval || CONST.APISERVER.WATCHER_REFRESH_INTERVAL}`);
```
you are right. Copy, paste issue.

jagadish-kb commented on api-controllers/FabrikBaseController.js:
```diff
@@ -35,12 +37,29 @@ class FabrikBaseController extends BaseController {
-      return this._lockResource(req, operationType)
+      let lockId, serviceFlowId, serviceFlowName;
+      return Promise.try(() => {
+          if (operationType === CONST.OPERATION_TYPE.UPDATE) {
```
yes.

jagadish-kb commented on operators/serviceflow-operator/SerialServiceFlowOperator.js:
```diff
@@ -0,0 +1,186 @@
+          CONST.APISERVER.RESOURCE_TYPES.TASK,
+          statesToWatchForTaskRelay,
+          (event) => this.relayTask(event),
+          CONST.APISERVER.POLLER_WATCHER_REFRESH_INTERVAL);
```
This is the poller watcher interval - This is similar to the BackupStatus Poller, refer [here](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/operators/BaseStatusPoller.js#L117) Pollers have different refresh interval.

ketakig12 replied:
Okay. Thanks!
